### PR TITLE
Kb rework

### PR DESCRIPTION
**Pull request by ea-rus**
## Description

Based on https://github.com/mindsdb/mindsdb/pull/8123

Changes for vector db:
- Changed interface of vector handlers:
  - don't return HandlerResponse (almost every where)
  - adapted chromadb and pgvector handlers
- support upsert in base class
   - if vector handler supports upsert - use internal method
   - otherwise do it using select/insert/update of vector handler  
- if 'id' column is not presented in input - used hash from content instead  
 
Reworked knowledge db (KB):
- request to KB routed through planner
  - it means that KB is a table and will receive queries which content only simple conditions, without subselects 
- Logic
  - if query is select: 
    - replace  CONTENT with EMBEDDINGS in conditions
    - send to vector db as query
  - if query is insert:
    - add EMBEDDINGS column to input data (based on CONTENT)
    - send to vector db as upsert
  - if query is update:
    - add  EMBEDDINGS column to updated columns
    - send to vector db as query

Fixes https://github.com/mindsdb/mindsdb_private/issues/235

Demo: https://www.loom.com/share/1a0aa01f107742a6a719be4467775de9

## Type of change

(Please delete options that are not relevant)

- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ⚡ New feature (non-breaking change which adds functionality)
- [x] 📢 Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] 📄 This change requires a documentation update

## Verification Process

To ensure the changes are working as expected:

 - [ ]   Test Location: Specify the URL or path for testing.
 - [ ]   Verification Steps: Outline the steps or queries needed to validate the change. Include any data, configurations, or actions required to reproduce or see the new functionality.

## Additional Media:

- [ ] I have attached a brief loom video or screenshots showcasing the new functionality or change.

## Checklist:

- [ ] My code follows the style guidelines(PEP 8) of MindsDB.
- [ ] I have appropriately commented on my code, especially in complex areas.
- [ ] Necessary documentation updates are either made or tracked in issues.
- [ ] Relevant unit and integration tests are updated or added.

### CODE REVIEW

**Review by dusvyat — APPROVED** [review]
lgtm

File: mindsdb/integrations/handlers/chromadb_handler/chromadb_handler.py [hunk 1]
```diff
@@ -120,8 +111,8 @@ def connect(self):
             self.is_connected = True
             return self._client
         except Exception as e:
+            log.logger.error(f"Error connecting to ChromaDB client, {e}!")
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/integrations/handlers/chromadb_handler/chromadb_handler.py:114]: ok, updated

File: mindsdb/api/mysql/mysql_proxy/datahub/datanodes/project_datanode.py [hunk 2]
```diff
@@ -63,58 +68,120 @@ def query(self, query=None, native_query=None, session=None):
...
+        from mindsdb.api.mysql.mysql_proxy.controllers.session_controller import SessionController
+        session = SessionController()
+
+        table_name = table_name.parts[-1]
+        kb_table = session.kb_controller.get_table(table_name, self.project.id)
+        if kb_table:
+            # this is the knowledge db
+            if is_replace:
+                # TODO clear table?
+                ...
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/api/mysql/mysql_proxy/datahub/datanodes/project_datanode.py:183]: added implementation

File: mindsdb/interfaces/knowledge_base/controller.py [hunk 3]
```diff
@@ -1,113 +1,283 @@
...
-        vector_database_id: str,
-        vector_database_table_name: str,
-        embedding_model_id: str,
+        project_name: str,
+        embedding_model: Identifier,
+        storage: Identifier,
         params: dict,
         if_not_exists: bool = False,
-    ) -> int:
+    ):
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/knowledge_base/controller.py:165]: added

File: mindsdb/interfaces/knowledge_base/controller.py [hunk 4]
```diff
@@ -1,113 +1,283 @@
...
 from mindsdb.integrations.libs.vectordatabase_handler import TableField
 
 
+class KnowledgeBaseTable:
+    def __init__(self, kb: db.KnowledgeBase, session):
+        self._kb = kb
+        self._vector_db = None
+        self.session = session
+
+    def select_query(self, query: Select):
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/knowledge_base/controller.py:27]: added

File: mindsdb/interfaces/knowledge_base/controller.py [hunk 5]
```diff
@@ -1,113 +1,283 @@
...
+    Star
 )
 
 import mindsdb.interfaces.storage.db as db
-from mindsdb.api.mysql.mysql_proxy.classes.sql_query import SQLQuery
-from mindsdb.api.mysql.mysql_proxy.executor.data_types import ANSWER_TYPE, ExecuteAnswer
 from mindsdb.integrations.libs.vectordatabase_handler import TableField
 
 
+class KnowledgeBaseTable:
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/knowledge_base/controller.py:21]: added

File: mindsdb/interfaces/knowledge_base/controller.py [hunk 6]
```diff
@@ -1,113 +1,283 @@
...
+
+        # not difference between cases in sql
+        name = name.lower()
+
+        kb = self.get(name, project_id)
+        if kb is not None:
+            if if_not_exists:
+                return kb
+            else:
+                raise Exception(f"Knowledge base already exists: {name}")
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/knowledge_base/controller.py:185]: changed

File: mindsdb/interfaces/knowledge_base/controller.py [hunk 7]
```diff
@@ -1,113 +1,283 @@
...
+
+        # model
+        model_name = embedding_model.parts[-1]
+
+        if len(embedding_model.parts) > 1:
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/knowledge_base/controller.py:190]: and what the improvement? :) 
no 'else' line in one case and no update model_project in another
they are the same options

File: mindsdb/interfaces/knowledge_base/controller.py [hunk 8]
```diff
@@ -1,113 +1,283 @@
...
+        project_id = project.id
+
+        # not difference between cases in sql
+        name = name.lower()
+
+        kb = self.get(name, project_id)
+        if kb is not None:
+            if if_not_exists:
+                return kb
+            else:
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/knowledge_base/controller.py:184]: removed

File: mindsdb/interfaces/knowledge_base/controller.py [hunk 9]
```diff
@@ -120,25 +290,12 @@ def get(self, name: str, project_id: str) -> db.KnowledgeBase:
...
-            db.session.query(db.KnowledgeBase)
-            .filter_by(
-                id=id,
-            )
-            .first()
-        )
-        if kb is None:
-            raise ValueError(f"Knowledge base not found: {id}")
-        return kb
+    def get_table(self, name: str, project_id: str):
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/knowledge_base/controller.py:295]: added

File: mindsdb/api/mysql/mysql_proxy/executor/executor_commands.py [hunk 10]
```diff
@@ -1293,100 +1280,35 @@ def answer_drop_view(self, statement):
...
             raise SqlApiException(
                 "Create a knowledge base from a select is not supported yet"
             )
 
-        params = statement.params
+        kb_name = statement.name.parts[-1]
 
         # create the knowledge base
-        _ = self.session.kb_controller.add(
+        self.session.kb_controller.add(
```
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/api/mysql/mysql_proxy/executor/executor_commands.py]: Returned, but for me it makes code a little bit dirtier visually. 
It is useful when you need to extract just part of the result (a, _ = function)
But I don't like to use it in any case when function returns something (I think it is 90% cases). 
This underscore just make us know that function returns something but don't inform what object and what type. To check this we have to look at the code or annotation anyway.

**Pull request merged**